### PR TITLE
EES- 4677 Create Build Pipeline

### PR DESCRIPTION
**Pull request by sambiramairelogic**
Creates two Azure DevOps pipelines:

* One suffixed `-pr` which should run automatically when changes are pushed to a branch that targets `main`. This pipeline builds each project, lints/type-checks, then runs all the tests projects. It will fail if any of these steps fail. 
* Another suffixed `-ci` which will run on the `main` branch whenever a change is pushed to it (i.e. when a PR is merged). This pipeline does not run linters or tests but does produce Artifacts which can then move on to infrastructure. 

In order to run the linters/tests, this PR also creates three very minimal skeleton test projects. It also makes a few configuration tweaks as and where necessary. 

Finally, I've taken the liberty of extending the vscode workspace file with a few extra folders and handy tasks. Very happy to amend/remove if there are any objections :) 

Pipelines in DevOps:

image (3)
image (2)
image (1)
image

### CODE REVIEW

File: infrastructure/azure-pipelines-pr.yaml [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+    - script: |
+        pip install pytest pytest-azurepipelines pytest-cov
+        pipenv run python -m pytest --cov=data_ingestion data_ingestion_tests/ --cov-report=xml --test-run-title="Data Ingestion API Tests"
+      displayName: 'Run and Publish Tests'
+
+    - task: PublishCodeCoverageResults@2
+      displayName: 'Collect Code Coverage'
+      inputs: 
+        summaryFileLocation: ./coverage.xml
```
Comment (benoutram) [inline infrastructure/azure-pipelines-pr.yaml:63]: I think before this we could run flake8 and black using 
`pipenv run black --check .` and `pipenv run flake8`. I think if both of those detect any differences in their checks then they will fail and should fail a pipeline task.

File: chatbot-ui/.eslintrc.json [hunk 2]
```diff
@@ -1,3 +1,21 @@
...
-  "extends": ["next/core-web-vitals", "plugin:prettier/recommended"]
+  "extends": [
+    "plugin:@typescript-eslint/recommended",
+    "plugin:react/recommended",
+    "plugin:react-hooks/recommended",
+    "plugin:jsx-a11y/recommended",
+    "plugin:import/typescript",
+    "prettier",
+    "prettier/prettier"
+  ],
```
Comment (benoutram) [inline chatbot-ui/.eslintrc.json:10]: I added "plugin:prettier/recommended" following the docs https://github.com/prettier/eslint-plugin-prettier to configure prettier within eslint. I'm just wondering with this change if prettier is still running with eslint. If not we might need to add a task into the pipeline to run a check with prettier.

File: .gitignore [hunk 3]
```diff
@@ -22,3 +22,5 @@ Thumbs.db
 chatbot-ui/.next/
 node_modules
 tsconfig.tsbuildinfo
+chatbot-ui/coverage/
+chatbot-ui/junit.xml
```
[the remark this answers is not in the document]
Comment (sambiramairelogic) [reply .gitignore:28]: Have added these above in the 'Python' section :)

File: chatbot-prototype.code-workspace [hunk 4]
```diff
@@ -1,16 +1,48 @@
...
+    },
     "search.exclude": {
       // Avoid polluting search results with lockfile content
       "Pipfile.lock": true,
       "pnpm-lock.yaml": true
     },
     // Ensure VSCode uses pnpm instead of npm
-    "npm.packageManager": "pnpm"
+    "npm.packageManager": "pnpm",
+    "jest.disabledWorkspaceFolders": [
```
[the remark this answers is not in the document]
Comment (sambiramairelogic) [reply chatbot-prototype.code-workspace:49]: Have added vscode-jest, python, and black as requested in the below extensions section. Also added docker and azure-pipelines tools too as these might be helpful

File: response_automater/config.py [hunk 5]
```diff
@@ -5,7 +5,7 @@ class Settings(BaseSettings):
     model_config = SettingsConfigDict(env_file=".env", extra="ignore")
 
     chat_url_public_ui: str = "http://localhost:3002"
-    openai_api_key: str
+    openai_api_key: str = "placeholder-value"
```
[the remark this answers is not in the document]
Comment (sambiramairelogic) [reply response_automater/config.py:8]: Good idea, have done

File: data_ingestion/utils.py [hunk 6]
```diff
@@ -7,7 +7,13 @@ def chunk_text(text):
     )
     chunks = []
     for record in text:
+        # is this working as intended?
+        # for record in "example" yields ['e', 'x', 'e'] etc. so record["text"] blows up
```
[the remark this answers is not in the document]
Comment (benoutram) [reply data_ingestion/utils.py:11]: Should be sorted by https://github.com/dfe-analytical-services/chatbot-prototype/pull/29

File: infrastructure/azure-pipelines-pr.yaml [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+      displayName: 'Run Flake8'
+      workingDirectory: data_ingestion
+
+    - script: |
+        pipenv run black ./ --check
+      displayName: 'Run Black'
+      workingDirectory: data_ingestion
+
+    - script: |
+        pip install pytest pytest-azurepipelines pytest-cov
```
Comment (benoutram) [inline infrastructure/azure-pipelines-pr.yaml:67]: I think we can get rid of this line

**Pull request merged**